servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java

```diff
@@ -893,14 +893,18 @@ public <R> R getEntries(
         }
       }
 
+      int filterDepth = namespaceDepth == null ? 0 : namespaceDepth.intValue();
       Predicate<KeyEntry> filterPredicate = filterEntries(filter);
+      if (filterDepth > 0) {
+        filterPredicate = filterPredicate.and(e -> e.getKey().elements().size() >= filterDepth);
+      }
 
       try (PaginationIterator<KeyEntry> entries =
           getStore()
               .getKeys(
                   refWithHash.getHash(),
                   pagingToken,
-                  withContent,
+                  withContent ? filterPredicate : null,
```

Comment:
I wonder if we shouldn't go one step further and merge `filterPredicate` with `KeyRestrictions.contentKeyPredicate`? The former operates on `KeyEntry`, the latter on `ContentKey`. But `KeyEntry` exposes `ContentKey` though the chain `KeyEntry.getKey().contentKey()`. So it should be possible to use just one predicate operating on `KeyEntry` and push it down to the Version Store.

Reply:
yes but it means you have to "materialize" a `KeyEntry` before you can apply the predicate... thus you'd need to do what i do in the PR i.e. build the simple `KeyEntry` first, and if the predicate matches, fetch the content and build a new special `KeyEntry` that also includes the content.

by merging those predicates we wouldnt have any immediate benefits in terms of preventing backend lookups + in-memory computations i think?

Reply:
I meant rather the opposite, i.e. the predicate you are adding is based on the namespace depth, and that can be expressed as a `Predicate<ContentKey>`. 

Therefore it should be possible to merge that predicate into `contentKeyPredicate` directly without even modifying `VersionStore.getKeys()`. Something like:

```java
      int filterDepth = namespaceDepth == null ? 0 : namespaceDepth.intValue();
      if (filterDepth > 0) {
        Predicate<ContentKey> depthFilter = e -> e.getElementCount() >= filterDepth;
        contentKeyPredicate = contentKeyPredicate == null ? depthFilter : contentKeyPredicate.and(depthFilter);
      }
```

That would prevent any key outside of the minimal depth to be even processed.

However, an open question remains: should we also push down the entire `filterPredicate` to avoid fetching unwanted contents? I think this is a good idea as the benefits of pushing the predicate down should vastly outweigh the cost of creating a `KeyEntry` instance. Except that this is a `Predicate<KeyEntry>` and as you pointed out, we'd need to construct the `KeyEntry` instance before applying the filter, which could potentially involve fetching the content. This is a chicken-and-egg problem.

For that part, I think we could get away by modifying `KeyEntry.getContent()` and making it a `@Lazy` attribute. We did something similar for `org.projectnessie.versioned.Put`, see https://github.com/projectnessie/nessie/blob/f8b9dceda1d12ba819f8fe8f7e9c831cae765ed7/versioned/spi/src/main/java/org/projectnessie/versioned/Put.java#L38-L41.

Reply:
I think there's still benefit in adding the `filter` to `KeyRestrictions.contentKeyPredicate`. This way, the `authz` iterator does not have to have extra checks and runs everything through the checkers (which is nicer from the security perspective).

We have to make a `IdentifiedContentKey` for each entry anyway, and we have to make a `KeyEntryForCel` on top of that if a filter is present. If we can refactor the code such that the extra work is not performed when the filter is not present, overall pushing the filter down seem beneficial to me.

Reply:
ah ok sorry for misunderstand, i've added the filter to `KeyRestrictions.contentKeyPredicate` but turns out `PersistVersionStore.getKeys` was not using that value at all (so tests started failing), so added a fix for that as well